server/src/main/java/org/opensearch/search/aggregations/metrics/SumAggregator.java

```diff
@@ -70,65 +71,58 @@ public class SumAggregator extends NumericMetricsAggregator.SingleValue {
         Map<String, Object> metadata
     ) throws IOException {
         super(name, context, parent, metadata);
-        // TODO: stop expecting nulls here
-        this.valuesSource = valuesSourceConfig.hasValues() ? (ValuesSource.Numeric) valuesSourceConfig.getValuesSource() : null;
+        this.valuesSource = (ValuesSource.Numeric) valuesSourceConfig.getValuesSource();
         this.format = valuesSourceConfig.format();
-        if (valuesSource != null) {
-            sums = context.bigArrays().newDoubleArray(1, true);
-            compensations = context.bigArrays().newDoubleArray(1, true);
-        }
+        sums = context.bigArrays().newDoubleArray(1, true);
     }
 
     @Override
     public ScoreMode scoreMode() {
-        return valuesSource != null && valuesSource.needsScores() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+        return valuesSource.needsScores() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
     }
 
     @Override
     public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, final LeafBucketCollector sub) throws IOException {
-        if (valuesSource == null) {
-            return LeafBucketCollector.NO_OP_COLLECTOR;
-        }
         final BigArrays bigArrays = context.bigArrays();
         final SortedNumericDoubleValues values = valuesSource.doubleValues(ctx);
-        final CompensatedSum kahanSummation = new CompensatedSum(0, 0);
         return new LeafBucketCollectorBase(sub, values) {
             @Override
             public void collect(int doc, long bucket) throws IOException {
                 sums = bigArrays.grow(sums, bucket + 1);
-                compensations = bigArrays.grow(compensations, bucket + 1);
 
                 if (values.advanceExact(doc)) {
                     final int valuesCount = values.docValueCount();
-                    // Compute the sum of double values with Kahan summation algorithm which is more
-                    // accurate than naive summation.
-                    double sum = sums.get(bucket);
-                    double compensation = compensations.get(bucket);
-                    kahanSummation.reset(sum, compensation);
+                    double value = sums.get(bucket);
```

Comment:
It sounds like the previous implementation returns completely wrong results though, which is a bug. I don't think we should trade off correctness for speed. Is there a path when `precise` can be chosen dynamically to correct those cases?

Reply:
> It sounds like the previous implementation returns completely wrong results though, which is a bug. 

@dblock this is not really a bug but implementation decision, it is well documented (on Elasticsearch [1] at least, we are lacking here) that sum is approximation:

> When running aggregations, Elasticsearch uses [double](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/number.html) values to hold and represent numeric data. As a result, aggregations on [long](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/number.html) numbers greater than 253 are approximate.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/7.10/search-aggregations.html#limits-for-long-values

Reply:
So that was surprising to me (and probably is surprising to users)! I think it does make sense to introduce `kahan` and `precise` options in that case and document the limitations of kahan. That said, is there a "bug where sum returns wrong values with long data type" then, or is it a known feature?

Reply:
> That said, is there a "bug where sum returns wrong values with long data type" then, or is it a known feature?

The behavior is documented - so I don't think this is a bug (but who reads the docs, right?). by the the docs - this is known behavior (or feature). The 100% bug for me at this point is that OpenSearch docs do not mention that.

> At which point is "approximation" actually a bug?

It is stated in docs as well - over certain range the approximation will happen (I would also emphasize that it will happen in any case but we could improve certainly flows to do math more accurately with this pull request).

Reply:
@matthewryanwells will you be able to help add/clarify docs and we can all go back to discussing the implementation here :)

Reply:
Whether it is technically a bug or not, I believe that by simply having different sum aggregators for floating point and integer values would allow the long values to be completely accurate and then we can change the documentation to match. This would have a minimal hit to performance because it just introduces a single function call, and would allow us to leave the kahan approximation working as is untouched .

Reply:
@dblock I just want to confirm that you are asking that I update the OpenSearch Aggregation documentation page to mention this limitation of the long data type, or are you suggesting I make an issue on the OpenSearch Github page to discuss this?

Reply:
If you just want to make the update that'd be awesome! At a minimum do open an issue so we don't forget. Appreciate it.

Reply:
> Whether it is technically a bug or not, I believe that by simply having different sum aggregators for floating point and integer values would allow the long values to be completely accurate and then we can change the documentation to match. This would have a minimal hit to performance because it just introduces a single function call, and would allow us to leave the kahan approximation working as is untouched.

I support this.

> Edit: I however have not considered considered how to fix the issue of longs easily overflowing if too many are added together.

`BigInteger`? ;)

Reply:
@dblock that is something I had just remembered existing and pushed a new commit to implement this, I have also just learned of the existence of feature flags which I could also possibly implement which I believe would allow users to enable or disable this more precise, but slower long aggregation. I will have to look into this further.

Reply:
@matthewryanwells just a note of this, feature flag are set on JVM level (at least for now this is the design decision), so the trade off here is clusters need to be restarted (vs aggregation parameter to control the behaviour at runtime).

Reply:
hmm, maybe first I will try to see if I can run performance tests to see if the performance hit from switching to `biginteger` is even that bad

Reply:
I have completed some rough performance tests by testing how much time and memory it takes to create and add 10,000 numbers to a long, and a BigInteger. In the end I found that BigInteger is roughly 25x slower and uses roughly 345,000,000x more memory (ouch).

Here are the statistics, keeping in mind this is the time and memory for 10,000 operations:
| Data Type  | Throughput | Allocation rate|
| ------------- | ------------- |------------- |
| BigInteger  | 12op/s | 13,801 MB/sec |
| Long  | 299op/s | 0.00004 MB/sec |

From these statistics I believe it is worth looking into a feature flag for this feature due to the large time and memory hit it causes.